File: persistence/scaladsl/src/main/scala/com/lightbend/lagom/scaladsl/persistence/AkkaTaggerAdapter.scala
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package com.lightbend.lagom.scaladsl.persistence

import akka.annotation.ApiMayChange
import akka.cluster.sharding.typed.scaladsl.EntityContext

@ApiMayChange
object AkkaTaggerAdapter {
  /**
   * Adapts an existing Lagom [[AggregateEventTagger]] to a
   * function {{{Event => Set[String]}}} as expected by Akka Persistence Typed {{{EventSourcedBehavior.withTagger}}} API.
   */
  def fromLagom[Command, Event <: AggregateEvent[Event]](
      entityCtx: EntityContext[Command],
      lagomTagger: AggregateEventTagger[Event]
  ): Event => Set[String] = { evt =>
    val tag =
      lagomTagger match {
        case tagger: AggregateEventTag[_] =>
          tagger.tag
        case shardedTagger: AggregateEventShards[_] =>
          shardedTagger.forEntityId(entityCtx.entityId).tag
      }
    Set(tag)
  }
}

File: persistence/javadsl/src/main/java/com/lightbend/lagom/javadsl/persistence/AkkaTaggerAdapter.java
/*
 * Copyright (C) Lightbend Inc. <https://www.lightbend.com>
 */

package com.lightbend.lagom.javadsl.persistence;

import akka.annotation.ApiMayChange;
import akka.cluster.sharding.typed.javadsl.EntityContext;
import java.util.HashSet;
import java.util.Set;
import java.util.function.Function;

@ApiMayChange
public class AkkaTaggerAdapter {

  /**
   * Adapts an existing Lagom {@code AggregateEventTagger} to a function {@code Function<Event,
   * Set<String>>} as expected by Akka Persistence Typed {@code EventSourcedBehavior.withTagger}
   * API.
   */
  public static <Command, Event extends AggregateEvent<Event>>
      Function<Event, Set<String>> fromLagom(
          EntityContext<Command> entityContext, AggregateEventTagger<Event> lagomTagger) {
    return evt -> {
      Set<String> tags = new HashSet<>();
      if (lagomTagger instanceof AggregateEventTag) {
        tags.add(((AggregateEventTag) lagomTagger).tag());
      } else if (lagomTagger instanceof AggregateEventShards) {
        tags.add(
            ((AggregateEventShards) lagomTagger).forEntityId(entityContext.getEntityId()).tag());
      }
      return tags;
    };
  }
}
